crates/kitsune_p2p/types/src/tx2/tx2_utils/latency.rs

```diff
@@ -0,0 +1,88 @@
+use once_cell::sync::Lazy;
+
+/// this is a reference instance in time
+/// all latency info is encoded relative to this
+static LOC_EPOCH: Lazy<std::time::Instant> = Lazy::new(std::time::Instant::now);
+
+/// this tag identifies that a latency marker will follow
+/// as a little endian ieee f64, this will decode to NaN.
+const LAT_TAG: &[u8; 8] = &[0xff, 0xff, 0xff, 0xfe, 0xfe, 0xff, 0xff, 0xff];
+
+/// Fill a buffer with data that is readable as latency information.
+/// Note, the minimum message size to get the timing data across is 16 bytes.
+pub fn fill_with_latency_info(buf: &mut [u8]) {
+    if buf.is_empty() {
+        return;
+    }
+
+    // make sure we call this first, so we don't go back in time
+    let epoch = *LOC_EPOCH;
+
+    let now = std::time::Instant::now();
+    let now = now.duration_since(epoch).as_secs_f64();
+
+    // create a pattern of tag/marker
+    let mut pat = [0_u8; 16];
+    pat[0..8].copy_from_slice(LAT_TAG);
+    pat[8..16].copy_from_slice(&now.to_le_bytes());
+
+    // copy the tag/marker pattern repeatedly into the buffer
+    let mut offset = 0;
+    while offset < buf.len() {
+        let len = std::cmp::min(pat.len(), buf.len() - offset);
+        buf[offset..offset + len].copy_from_slice(&pat[..len]);
+        offset += len;
+    }
+}
+
+/// Return the duration since the time encoded in a latency info buffer.
+/// Returns a unit error if we could not parse the buffer into time data.
+pub fn parse_latency_info(buf: &[u8]) -> Result<std::time::Duration, ()> {
```

Comment:
What is the actual usecase for this? I don't understand why the entire buffer is filled with the same timestamp?

Reply:
This lets me do things like set up an AsyncRead trait object that just spits out infinite latency timing data, so when I read from it, and do other networking operations with that data, at the final position I can parse the data and see how long it took to get to the end.

I should make the writer compatible with PoolBuf though, so I don't have to initialize the slice first before filling it with the latency data...